File: others/contains_duplicate_3.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool containsNearbyAlmostDuplicate(vector<int>& nums, int k, int t) {
        set<long long> prev,nxt;
        for(int i=0;i<nums.size();i++){
            long long cur=nums[i];
            auto pit = prev.lower_bound(-1*cur); 
            auto nit = nxt.lower_bound(cur);
            if(pit!=prev.end() && (*pit)+cur<=1LL*t)    return true;
            if(nit!=nxt.end() && (*nit)-cur<=1LL*t)    return true;

            prev.insert(-1*cur);
            nxt.insert(cur);
            if(prev.size()>k)   prev.erase(-1*nums[i-k]);
            if(nxt.size()>k)   nxt.erase(nums[i-k]);
        }
        return false; 
    }
};

File: others/contains_duplicate_3.py
from collections import defaultdict
class Solution:
    def containsNearbyAlmostDuplicate(self, nums: List[int], k: int, t: int) -> bool:
        if t<0:
           return False
        hashmap = defaultdict(list)
        for i in range(len(nums)):
            key = nums[i]//(t+1)
            if hashmap[key]:
                return True
            elif hashmap[key-1] and abs(nums[i]-hashmap[key-1][0])<=t:
                return True
            elif hashmap[key+1] and abs(nums[i]-hashmap[key+1][0])<=t:
                return True
            hashmap[key].append(nums[i])
            if i>=k:
                del_key = nums[i-k]//(t+1)
                hashmap[del_key].pop(0)
        return False
